packages/flutter/lib/src/material/scaffold.dart

```diff
@@ -275,6 +275,79 @@ class _ScaffoldGeometryNotifier extends ChangeNotifier implements ValueListenabl
   }
 }
 
+// Used to communicate the height of the Scaffold's bottomNavigationBar and
+// persistentFooterButtons to the LayoutBuilder which builds the Scaffold's body.
+//
+// Scaffold expects a _BodyBoxConstraints to be passed to the _BodyBuilder
+// widget's LayoutBuilder, see _ScaffoldLayout.performLayout(). The BoxConstraints
+// methods that construct new BoxConstraints objects, like copyWith() have not
+// been overridden here because we expect the _BodyBoxConstraintsObject to be
+// passed along unmodified to the LayoutBuilder. If that changes in the future
+// then _BodyBuilder will assert.
+class _BodyBoxConstraints extends BoxConstraints {
+  const _BodyBoxConstraints({
+    double minWidth = 0.0,
+    double maxWidth = double.infinity,
+    double minHeight = 0.0,
+    double maxHeight = double.infinity,
+    @required this.bottomWidgetsHeight,
+  }) :  assert(bottomWidgetsHeight != null),
+        assert(bottomWidgetsHeight >= 0),
+        super(minWidth: minWidth, maxWidth: maxWidth, minHeight: minHeight, maxHeight: maxHeight);
+
+  final double bottomWidgetsHeight;
+
+  // RenderObject.layout() will only short-circuit its call to its performLayout
+  // method if the new layout constraints are not == to the current constraints.
+  // If the height of the bottom widgets has changed, even though the constraints'
+  // min and max values have not, we still want performLayout to happen.
+  @override
+  bool operator ==(dynamic other) {
+    if (super != other)
+      return false;
+    if (other is! _BodyBoxConstraints)
+      return false;
+    final _BodyBoxConstraints typedOther = other;
+    return bottomWidgetsHeight == typedOther.bottomWidgetsHeight;
+  }
+
+  @override
+  int get hashCode {
+    assert(debugAssertIsValid());
+    return hashValues(minWidth, maxWidth, minHeight, maxHeight, bottomWidgetsHeight);
+  }
+}
+
+// Used when Scaffold.extendBody is true to wrap the scaffold's body in a MediaQuery
+// whose padding accounts for the height of the bottomNavigationBar and/or the
+// persistentFooterButtons.
+//
+// The bottom widgets' height is passed along via the _BodyBoxConstraints parameter.
+// The constraints parameter is constructed in_ScaffoldLayout.performLayout().
+class _BodyBuilder extends StatelessWidget {
+  const _BodyBuilder({ Key key, this.body }) : super(key: key);
+
+  final Widget body;
+
+  @override
+  Widget build(BuildContext context) {
+    return LayoutBuilder(
+      builder: (BuildContext context, BoxConstraints constraints) {
+        final _BodyBoxConstraints bodyConstraints = constraints;
+        if (bodyConstraints.bottomWidgetsHeight == 0.0)
+          return body;
+        final MediaQueryData metrics = MediaQuery.of(context);
+        return MediaQuery(
+          data: metrics.copyWith(
+            padding: metrics.padding.copyWith(bottom: bodyConstraints.bottomWidgetsHeight),
```

Comment:
do we guarantee that `bodyConstraints.bottomWidgetsHeight > metrics.padding.bottom` ? If not, we should use `math.max` to make it the bigger of `bodyConstraints.bottomWidgetsHeight` or `metrics.padding.bottom`.

Reply:
I've added the `math.max`, that should be safest.